### DISCUSSION TITLE
What's the Phalcon 5 roadmap?

### DISCUSSION THREAD

**Discussion by zz** [phalcon/phalcon#86] [Q&A]
What's the Phalcon 5 roadmap?

**Comment by niden** [comment]
We have not finalized it per se on a document, it has only been discussed. Just one more todo in our list

In short:

- Support v4 with security fixes etc. No new features. 
- Try and get v4 in PHP 8 (ongoing)
- _Translate_ (refactor) v4 from Zephir to PHP -> v5
- Minimal changes if any from v4 to v5
- Get v5 to composer
- Add metrics to v5 to see how slow it is compared to the extension

After that, we will switch from the MVC pattern to the ADR one, making Phalcon as slim and customizable as possible while keeping performance as fast as possible. But that is quite a long way, we still have to release v5 and that is quite a lot of work.

Of course an extension will always be much faster than PHP code and it will use less memory. We are however introducing benchmarks to ensure that any change we make, speeds up the framework.

**Reply by madrian-es** [reply]
What is the reason for moving from MVC to ADR?

**Comment by AliShareei** [comment]
Hi @niden,

If we create our application with Phalcon V4, we can expect after releasing V5, it's quite working too?
I mean you think we can work with V4 but our goal be for v5?

Because I'm trying to use `composer require phalcon/phalcon:v5.0.x-dev`, but it' seems some class like `Phalcon\Mvc\Application` does not exist at this time.

Do you think when V5 will be released, with all part of v4? 

Thanks.

**Comment by niden** [comment]
@AliShareei v5 is not ready yet - you can see it really, classes are missing for now.

The goal is to have v5 as identical to v4 as possible. We are aiming for a drop in replacement or at least very very minimal changes from v4 to v5. v5 will contain everything v4 has already :)

**Comment by AliShareei** [comment]
@niden, I think it's a good idea to make compatible these two different versions. 
if we create a project with v4 I guess we can even use it in shared hosting with v5.

Do you think we can have v5 until May? I mean within 4-5 months later?

**Reply by niden** [reply]
> @niden, I think it's a good idea to make compatible these two different versions.
> if we create a project with v4 I guess we can even use it in shared hosting with v5.
> 
> Do you think we can have v5 until May? I mean within 4-5 months later?

This is where we are with v5 so far: https://github.com/orgs/phalcon/projects/5

Still quite a few cards to do but the ones that are done, the tests are passing and the changes are minimal if any for each component/namespace

**Comment by Jeckerson** [comment]
> Do you think we can have v5 until May? I mean within 4-5 months later?

Quite possible.

**Comment by Attax** [comment]
How much loss will we  got when we move to a pure php framework  from a c extension ？The high performance and friendly restful api route are the most important reasons that why we choose phalcon. Maybe we should choose other C extension library if we got too much performance loss，evenif the other library has a very unfriendly route.

**Reply by niden** [reply]
The loss of performance is unknown at this point. The work in progress at the moment for v6 (pure PHP) will include a set of benchmarks for the framework to ensure that we have a measure of how fast the framework is. We will introduce those benchmarks on the v5 version (extension) and see where we have the highest loss of performance.

After that, we will also introduce an extension that will "override" the php based classes or methods when the extension is loaded. This has been successfully tested and works in the Tensor library. If the extension is not loaded, then the PHP code continues to run, otherwise it uses the extension. This way we can have people that want this extra performance install the supplementary extension and get a higher performance from the framework.

**Reply by Attax** [reply]
thank you!

**Comment by RiteshAryal** [comment]
Is Phalcon v5 ready for Production use, using PHP8.x ? Please suggest with few relevant links.
Appreciate your help.

**Reply by sinbadxiii** [reply]
My personal opinion: v5 is quite ready for use in production, one of the e-commerce shop that uses Phalcon v5 RC3 - https://ankas.ru

**Reply by RiteshAryal** [reply]
Which PHPx version is it using ?

**Reply by sinbadxiii** [reply]
> Which PHPx version is it using ?

PHP 8.1

**Comment by andypost** [comment]
Is compatibility with PHP 8.2 will be delayed til 5,1?

**Reply by Jeckerson** [reply]
Depends how much changes there will be in Zephir.
